proposals/simd-0170-builtin-instruction-cost-and-budget.md

```diff
@@ -0,0 +1,106 @@
+allocates a much larger value (DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT) to each
+built-in instruction. This discrepancy creates imprecise CU tracking and may
+require block producers to account for additional CUs after a transaction's
+execution, potentially causing block limits to be exceeded.
+
+   Furthermore, built-ins can conditionally CPI into other programs, introducing
+variability in CU consumption. This unpredictability makes it difficult to
+allocate CUs upfront solely based on CUs builtin consumes during execution, also
+adds complexity to the tracking process.
+
+2. **Preventing over-allocation of CUs**: Over-allocating CUs for built-ins
+   reduces block density, lowers network throughput, and can degrade block
+producer performance by causing repeated transaction retries. Avoiding excessive
+CU allocation is critical for maximizing block efficiency and minimizing network
+delays.
+
+To resolve these issues, we propose statically defining the CU cost of built-in
+instructions based on worst-case scenarios, including potential CPI calls. A
+unified MAX_BUILTIN_ALLOCATION_COMPUTE_UNIT_LIMIT will standardize CU allocation
+for both block producers and the CU meter.
+
+## Alternatives Considered
+
+1. **Maintain current CU allocation with additional tracking logic**: One option
+   is to keep the current DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT for built-ins
+and add/modify tracking logic to the cost tracker. This would address the
+discrepancy between CU allocation and consumption but increase system
+complexity. The added logic could introduce corner cases and potential bugs,
+raising the risk of issues in the transaction pipeline.
+
+2. **Treat built-ins like regular instructions**: Another approach would be to
+   allocate DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT for built-ins as for other
+instructions. However, this fails to address over-allocation. Unlike regular
+instructions, the execution of built-ins is more predictable and usually
+requires significantly fewer CUs than BPF instructions. This approach would
+allocate more CUs than necessary, undermining the goal of efficient CU usage.
+
+3. **Declare both max and default CU values for built-ins**: A more precise
+   approach would be to require built-ins to declare both a maximum CU
+allocation (MAX_BUILTIN_ALLOCATION_COMPUTE_UNIT_LIMIT) and a default CU value
+for each instruction (DEFAULT_BUILTIN_INSTRUCTION_COMPUTE_UNITS). This would
+allow fine-tuned CU allocation based on each instruction’s potential execution
+path. However, this introduces new constraints for existing and future
+built-ins, requiring updates to comply with these rules, which could
+overcomplicate the design.
+
+## Detailed Design
+
+1. **Statically define CUs per instruction**: Assign a fixed CU consumption
+   (DEFAULT_BUILTIN_INSTRUCTION_COMPUTE_UNITS) to each built-in instruction
+rather than per built-in program.
+2. **Set a static maximum CU allocation**: Propose a global limit of 5,000 CUs
```

Comment:
`5000` is based on existing builtins. AddressLookupTable's `CreateLookupTable` makes most CPI calls (up to 3 calls to `System`), that results up to `750 + 3*150 = 1200 CU`; Upgradeable Loader's `ExtendProgram` may CPI `System` once, in that case it'd need `2370 + 150 = 2520 CU`, which is the highest of all existing builtins. `5000` nearly doubles that, to leave room for future builtins.

Reply:
Is the `5000` figure enough to increase block density by an adequately significant margin? Or should it be tighter? Seems like this would be a pretty dramatic improvement, right?

Reply:
Yes, the `5,000` yields a good result so far. 

In testing, I looked at 3 metrics: 
1. Should have zero case of "adjust-up" at producer side, and
2. Produced blocks should be (nearly) full, and
3. Count of transaction retries (due to "would-exceed-limits") should not be too high (indicating not over allocating);

Compare to original `200,000`, setting builtin allocation to `5000` on both producer and CU meter have a good result: no adjust-up, blocks are always (near) full, and number of retries were up very slightly. 

Setting it tighter to `3,000` didn't make too much improvements from `5,000`.